### ISSUE TITLE
Structure loss

### ISSUE THREAD

**Issue by mememimis** [ruotianluo/self-critical.pytorch#152]
Hi,

I have a question regarding the structure loss in su+bu+structure. The structure loss weight is similar as the lambda in your discriminality paper?  That  is scales the rewards with  cross-entropy ? Furthermore, if that is the case each of the rewards can as well have different weights? 

Thank you.

**Comment by mememimis** [comment]
Additionally, the cider scores is too low : 0.066 while in self-critical works fine

**Comment by ruotianluo** [comment]
It's not the same. That is the weighting between structure and cross entropy. You can think of structure loss as an alternative to self critical.

**Comment by mememimis** [comment]
And it is in effect when  train for multiple  captions?

**Comment by mememimis** [comment]
When   self_critical_reward_weight=1 then is balancing  cider and cross entropy?

**Comment by ruotianluo** [comment]
Yes. 
0.5 is balance

**Comment by mememimis** [comment]
Thank you for your kind  help. This part is very interesting.

**Comment by mememimis** [comment]
I have two more questions , the structure loss is a bit unclear to me:
1) When the weight of structure loss is 1 it is in effect?
2) structure_n_sample why it is set as 16?
3) can cider reward be used with entropy  when entropy reward is set to 1?
4) self-cider can be used with cider, as blue?

Finally, when sampling diverse beam search should be 1 and sample_n should be 3 when we want to sample a set of 3? 

Thank you very much.

**Comment by ruotianluo** [comment]
1 dont understsnf
2 it's from the structure loss paper
3 yes. But from my experience, the entropy reward doesn't do much. 
4 technically yes. But I tried, doesnt do much either.

**Comment by mememimis** [comment]
Could you please mention the title of the structure loss paper?

**Comment by ruotianluo** [comment]
Classical Structured Prediction Losses for Sequence to Sequence Learning (Edunov et al., 2018)

**Comment by mememimis** [comment]
Fantastic! Thank you! It is of great help to understand this part of the
code.

**Comment by mememimis** [comment]
When im training with any of the structure losses I'm getting the following :  train_loss = 7.562, **lm_loss = 0.000**, struc_loss = 1.512 . Should lm_loss be 0?

**Comment by ruotianluo** [comment]
Sorry, forgot to look into it yesterday.  Was there a bug?

**Comment by mememimis** [comment]
No worries, i still haven't figure it out yet.  The loss it always zero. It is the case for you as well or i have to look into the versions of what i'm using etc? 
I have another question as well.  The urnng that the new self critical is from is the : Unsupervised Recurrent Neural Network Grammars?

**Comment by ruotianluo** [comment]
In fact it's from Variationalinference for monte carlo objectives. But I saw this first time in urnng.

**Comment by mememimis** [comment]
i think is the other way around,

`            if opt.structure_loss_weight < 1:
                lm_loss = torch.tensor(0).type_as(fc_feats)

            else:
                lm_loss = self.crit(self.model(fc_feats, att_feats, labels, att_masks), labels[:,1:], masks[:,1:])
`

it should be like instead the oposite.

**Comment by ruotianluo** [comment]
The way I am using is structure_loss_weight should be between 0 and 1, so >1 means =1 means no lm_loss

**Comment by mememimis** [comment]
A thank you  very much  didn't  thought of it that way!  I thought values between  0-1  were disregarded. Thank you!!

**Comment by mememimis** [comment]
I encountered  a problem regarding the structure loss when the type is risk or  softmax_margin after a couple of  batches the structure loss becomes nan along with the train_loss and it breaks:
train_loss = 1.763, lm_loss = 2.209, struc_loss = 0.723
train_loss = nan, lm_loss = 2.159, struc_loss = nan
have you encounter it as well?

**Comment by ruotianluo** [comment]
I am not sure. The structure loss is experimental, and I think I didn't make all of them work. You may need to explore and tweak the training for example learning rate etc by yourself.
